src/zalsa.rs

```diff
@@ -296,29 +296,46 @@ impl Zalsa {
             .expect("should have the ingredient index available")
     }
 
+    /// **NOT SEMVER STABLE**
+    #[doc(hidden)]
+    #[inline]
+    pub fn lookup_jar_by_type<J: Jar>(&self) -> Option<IngredientIndex> {
+        let jar_type_id = TypeId::of::<J>();
+        self.jar_map.get(&jar_type_id, &self.jar_map.guard())
+    }
```

Comment:
This is a bit unfortunate that we need a new method for this. I guess it shouldn't matter in the fast-path where we always exit early but it does mean that `self.jar_map.get` is called three times in a slow path of a query (and that might affect inlining). 

Could we rewrite this to an entry-style API where `jar_by_type` returns an `Entry` struct that has an `is_add` function (or similar) and a `get` to get the value (which might insert it).